cas-client-core/src/main/java/org/jasig/cas/client/validation/json/TicketValidationJsonResponse.java

```diff
@@ -0,0 +1,135 @@
+package org.jasig.cas.client.validation.json;
+
+import org.jasig.cas.client.authentication.AttributePrincipal;
+import org.jasig.cas.client.authentication.AttributePrincipalImpl;
+import org.jasig.cas.client.proxy.ProxyGrantingTicketStorage;
+import org.jasig.cas.client.proxy.ProxyRetriever;
+import org.jasig.cas.client.util.CommonUtils;
+import org.jasig.cas.client.validation.Assertion;
+import org.jasig.cas.client.validation.AssertionImpl;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This is {@link TicketValidationJsonResponse}.
+ *
+ * @author Misagh Moayyed
+ */
+final class TicketValidationJsonResponse {
+    private CasServiceResponseAuthentication serviceResponse;
+
+    public void setServiceResponse(final CasServiceResponseAuthentication serviceResponse) {
+        this.serviceResponse = serviceResponse;
+    }
+
+    public CasServiceResponseAuthentication getServiceResponse() {
+        return serviceResponse;
+    }
+
+    Assertion getAssertion(final ProxyGrantingTicketStorage proxyGrantingTicketStorage,
+                           final ProxyRetriever proxyRetriever) {
+        final String proxyGrantingTicketIou = getServiceResponse().getAuthenticationSuccess().getProxyGrantingTicket();
+        final String proxyGrantingTicket;
+        if (CommonUtils.isBlank(proxyGrantingTicketIou) || proxyGrantingTicketStorage == null) {
+            proxyGrantingTicket = null;
+        } else {
+            proxyGrantingTicket = proxyGrantingTicketStorage.retrieve(proxyGrantingTicketIou);
+        }
+
+        final Assertion assertion;
+        final Map<String, Object> attributes = getServiceResponse().getAuthenticationSuccess().getAttributes();
+        final String principal = getServiceResponse().getAuthenticationSuccess().getUser();
+        if (CommonUtils.isNotBlank(proxyGrantingTicket)) {
+            final AttributePrincipal attributePrincipal = new AttributePrincipalImpl(principal, attributes,
+                    proxyGrantingTicket, proxyRetriever);
+            assertion = new AssertionImpl(attributePrincipal);
+        } else {
+            assertion = new AssertionImpl(new AttributePrincipalImpl(principal, attributes));
+        }
+        return assertion;
+    }
+
+    static class CasServiceResponseAuthentication {
+        private CasServiceResponseAuthenticationFailure authenticationFailure;
+        private CasServiceResponseAuthenticationSuccess authenticationSuccess;
```

Comment:
It strikes me as odd that there are slots for both success and failure, but if that's what the protocol specifies I suppose it's necessary. I recall seeing checks in the validator that both fields are not set simultaneously, which is the concern in practice, so I suppose it's all good. If the JSON parser didn't impose requirements on the JSON structure, making the fields final and setting them in the constructor would be better.